robot/impl/resource_manager.go

```diff
@@ -529,6 +530,152 @@ func (manager *resourceManager) completeConfig(
+						fmt.Errorf("resource config validation error: %w", err),
+						"resource", conf.ResourceName(),
+						"model", conf.Model)
+					return
+				}
+				if manager.moduleManager.Provides(conf) {
+					if _, err := manager.moduleManager.ValidateConfig(ctxWithTimeout, conf); err != nil {
+						gNode.LogAndSetLastError(
+							fmt.Errorf("modular resource config validation error: %w", err),
+							"resource", conf.ResourceName(),
+							"model", conf.Model)
+						return
+					}
+				}
+
+				switch {
+				case resName.API.IsComponent(), resName.API.IsService():
+					processResource := func() {
+						defer levelWG.Done()
+
+						newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
+						if newlyBuilt || err != nil {
+							if err := manager.markChildrenForUpdate(resName); err != nil {
+								manager.logger.CErrorw(ctx,
+									"failed to mark children of resource for update",
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+					}
+
+					if !forceSync {
+						// TODO(RSDK-6925): support concurrent processing of resources of
+						// APIs with a maximum instance limit. Currently this limit is
+						// validated later in the resource creation flow and assumes that
+						// each resource is created synchronously to have an accurate
+						// creation count.
+						if c, ok := resource.LookupGenericAPIRegistration(resName.API); ok && c.MaxInstance != 0 {
+							forceSync = true
+						}
+					}
+
+					levelWG.Add(1)
+					if forceSync {
+						processResource()
+					} else {
+						go processResource()
+					}
+
+				default:
+					err := errors.New("config is not for a component or service")
+					gNode.LogAndSetLastError(err, "resource", resName)
+				}
+			})
+
+			select {
+			case <-resChan:
+			case <-ctxWithTimeout.Done():
+				if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+					lr.logger.CWarn(ctx, rutils.NewBuildTimeoutError(resName.String()))
+				}
+			case <-ctx.Done():
+				return
+			}
+		} // for-each resource name
+		levelWG.Wait()
+	} // for-each level
+}
+
+func (manager *resourceManager) completeConfigForRemotes(ctx context.Context, lr *localRobot) {
```

Comment:
Can you confirm @maximpertsov this was just a copy-paste code refactor?

Reply:
yes, no changes aside from renaming the `*localRobot` variable to `lr`. I'm also happy to undo this change to make the PR less noisy